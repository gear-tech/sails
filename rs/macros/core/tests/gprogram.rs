use proc_macro2::TokenStream;
use quote::quote;
use sails_macros_core::__gprogram_internal as gprogram;

#[test]
fn generates_init_for_single_ctor() {
    let input = quote! {
        impl MyProgram {
            pub async fn new(p1: u32, p2: String) -> Self {
                Self { p1, p2 }
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_init_for_multiple_ctors() {
    let input = quote! {
        impl MyProgram {
            pub async fn new(p1: u32, p2: String) -> Self {
                Self { p1, p2 }
            }

            pub fn new2(p2: String, p1: u32) -> Self {
                Self { p1, p2 }
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_init_for_no_ctor() {
    let input = quote! {
        impl MyProgram {
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_for_single_service_with_non_empty_route() {
    let input = quote! {
        impl MyProgram {
            pub fn service(&self) -> MyService {
                MyService
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_for_multiple_services_with_non_empty_routes() {
    let input = quote! {
        impl MyProgram {
            #[route("svc1")]
            pub fn service1(&self) -> MyService {
                MyService
            }

            pub fn service2(&self) -> MyService {
                MyService
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_with_gprogram_attributes() {
    let args = quote!(
        handle_reply = my_handle_reply,
        handle_signal = my_handle_signal
    );
    let input = quote! {
        impl MyProgram {}
    };

    let result = gprogram(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_with_crate_path() {
    let args = quote!(crate = sails_rename,);
    let input = quote! {
        impl MyProgram {}
    };

    let result = gprogram(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
