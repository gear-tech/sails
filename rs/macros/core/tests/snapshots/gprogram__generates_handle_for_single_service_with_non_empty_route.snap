---
source: rs/macros/core/tests/gprogram.rs
expression: result
---
static __ROUTE_SERVICE: [u8; 8usize] = [
    28u8,
    83u8,
    101u8,
    114u8,
    118u8,
    105u8,
    99u8,
    101u8,
];
impl MyProgram {
    pub fn service(&self) -> <MyService as sails_rs::gstd::services::Service>::Exposure {
        let service = self.__service();
        let exposure = <MyService as sails_rs::gstd::services::Service>::expose(
            service,
            sails_rs::gstd::msg::id().into(),
            __ROUTE_SERVICE.as_ref(),
        );
        exposure
    }
    fn __service(&self) -> MyService {
        MyService
    }
    pub fn default() -> Self {
        Self
    }
}
impl sails_rs::meta::ProgramMeta for MyProgram {
    fn constructors() -> sails_rs::scale_info::MetaType {
        sails_rs::scale_info::MetaType::new::<meta_in_program::ConstructorsMeta>()
    }
    fn services() -> impl Iterator<
        Item = (&'static str, sails_rs::meta::AnyServiceMeta),
    > {
        [("Service", sails_rs::meta::AnyServiceMeta::new::<MyService>())].into_iter()
    }
}
use sails_rs::Decode as __ProgramDecode;
use sails_rs::TypeInfo as __ProgramTypeInfo;
#[derive(__ProgramDecode, __ProgramTypeInfo)]
#[codec(crate = sails_rs::scale_codec)]
#[scale_info(crate = sails_rs::scale_info)]
#[allow(dead_code)]
struct __DefaultParams {}
mod meta_in_program {
    use super::*;
    #[derive(__ProgramTypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum ConstructorsMeta {
        Default(__DefaultParams),
    }
}
#[cfg(target_arch = "wasm32")]
pub mod wasm {
    use super::*;
    use sails_rs::{gstd, hex, prelude::*};
    static mut PROGRAM: Option<MyProgram> = None;
    #[gstd::async_init]
    async fn init() {
        sails_rs::gstd::events::__enable_events();
        let mut input: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
        let (program, invocation_route) = if input
            .starts_with(&[28u8, 68u8, 101u8, 102u8, 97u8, 117u8, 108u8, 116u8])
        {
            static INVOCATION_ROUTE: [u8; 8usize] = [
                28u8,
                68u8,
                101u8,
                102u8,
                97u8,
                117u8,
                108u8,
                116u8,
            ];
            let request = __DefaultParams::decode(&mut &input[8usize..])
                .expect("Failed to decode request");
            let program = MyProgram::default();
            (program, INVOCATION_ROUTE.as_ref())
        } else {
            let mut __input = input;
            let input = String::decode(&mut __input)
                .unwrap_or_else(|_| {
                    if input.len() <= 8 {
                        format!("0x{}", sails_rs::hex::encode(input))
                    } else {
                        format!(
                            "0x{}..{}", sails_rs::hex::encode(& input[..4]),
                            sails_rs::hex::encode(& input[input.len() - 4..])
                        )
                    }
                });
            panic!("Unexpected ctor: {}", input)
        };
        unsafe {
            PROGRAM = Some(program);
        }
        gstd::msg::reply_bytes(invocation_route, 0).expect("Failed to send output");
    }
    #[gstd::async_main]
    async fn main() {
        let mut input: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
        let output: Vec<u8> = if input.starts_with(&__ROUTE_SERVICE) {
            let program_ref = unsafe { PROGRAM.as_ref() }
                .expect("Program not initialized");
            let mut service = program_ref.service();
            let output = service.handle(&input[__ROUTE_SERVICE.len()..]).await;
            [__ROUTE_SERVICE.as_ref(), &output].concat()
        } else {
            let mut __input = input;
            let input = String::decode(&mut __input)
                .unwrap_or_else(|_| {
                    if input.len() <= 8 {
                        format!("0x{}", sails_rs::hex::encode(input))
                    } else {
                        format!(
                            "0x{}..{}", sails_rs::hex::encode(& input[..4]),
                            sails_rs::hex::encode(& input[input.len() - 4..])
                        )
                    }
                });
            panic!("Unexpected service: {}", input)
        };
        gstd::msg::reply_bytes(output, 0).expect("Failed to send output");
    }
}
